lib/ramble/ramble/expander.py

```diff
@@ -25,9 +25,203 @@
+        else:
+            self.children.append(children)
+
+    def define_value(self, expansion_dict, allow_passthrough=True,
+                     expansion_func=str, evaluation_func=eval):
+        """Define the value for this node.
+
+        Construct the value of self. This builds up a string representation of
+        self, and performs evaluation and formatting of the resulting string.
+        This includes extracting the values of the children nodes, and
+        replacing their values in the proper positions in self's string.
+
+        Stores the resulting value in self.value
+
+        Args:
+            expansion_dict (dict): variable definitions to use for expanding
+            detected matches
+            allow_passthrough (bool): if true, expansion is allowed to fail. if
+            false, failed expansion raises an error.
+            expansion_func (func): function to use for expansion of nested
+            variable definitions
+            evaluation_func (func): function to use for evaluating math of strings
+        """
+        if self.contents is not None:
+            parts = []
+            last_idx = 0
+            for child in self.children:
+                child_indices = child.relative_indices(self)
+                parts.append(self.contents[last_idx:child_indices[0]])
+                parts.append(str(child.value))
+                last_idx = child_indices[1] + 1
+
+            if last_idx != len(self.contents):
+                parts.append(self.contents[last_idx:])
+
+            if self != self.root:
+                replaced_contents = ''.join(parts)
+
+                # Special case '{}'
+                if len(replaced_contents) == 2:
+                    self.value = '{}'
+                    return
+
+                format_kw = replaced_contents[1:-1]
+                kw_parts = format_kw.split(':')
+                required_passthrough = False
+
+                if kw_parts[0] in expansion_dict:
+                    self.value = expansion_func(expansion_dict,
+                                                expansion_dict[kw_parts[0]],
+                                                allow_passthrough=allow_passthrough)
+                else:
+                    self.value = kw_parts[0]
+                    required_passthrough = True
+
+                # Evaluation should go here
+                try:
+                    old_value = self.value
+                    self.value = evaluation_func(self.value)
+                    if old_value != self.value:
+                        required_passthrough = False
+                except SyntaxError:
+                    pass
+
+                # If we had a format spec, add it
+                if len(kw_parts) > 1:
+                    kw_dict = {'value': self.value}
+                    format_str = f'value:{kw_parts[1]}'
+                    try:
+                        self.value = formatter.vformat('{' + format_str + '}', [], kw_dict)
```

Comment:
It would be nice to DRY out the literal `'{'`, especially in this file we kind of use two different delims with the same value?